packages/search/provider.tsx

```diff
@@ -0,0 +1,113 @@
+}
+
+export function useSearch(): SearchContextObject {
+  return useContext(SearchContext)
+}
+
+export default function SearchProvider({
+  children,
+  algoliaConfig = {},
+}: {
+  children: React.ReactNode
+  algoliaConfig: AlgoliaConfigObject
+}): React.ReactElement {
+  const [query, setQuery] = useState('')
+  const [isCancelled, setCancelled] = useState(false)
+
+  const [appId, indexName, apiKey] = [
+    algoliaConfig.appId || process.env.NEXT_PUBLIC_ALGOLIA_APP_ID,
+    algoliaConfig.indexName || process.env.NEXT_PUBLIC_ALGOLIA_INDEX,
+    algoliaConfig.searchOnlyApiKey ||
+      process.env.NEXT_PUBLIC_ALGOLIA_SEARCH_ONLY_API_KEY,
+  ]
+
+  if (!appId || !apiKey || !indexName) {
+    throw new Error(`Missing required Algolia arguments or .env variables for SearchProvider. If using environment variables, ensure the following are present in your environment:
+- NEXT_PUBLIC_ALGOLIA_APP_ID
+- NEXT_PUBLIC_ALGOLIA_SEARCH_ONLY_API_KEY
+- NEXT_PUBLIC_ALGOLIA_INDEX
+If passing the algoliaConfig prop to SearchProvider, ensure the following keys are defined:
+- appId
+- indexName
+- searchOnlyApiKey
+`)
+  }
+
+  // Memo-ize client context values, as these should only
+  // change and cause re-renders when Algolia configuration changes
+  const { client, initAlgoliaInsights, logClick } = useMemo(() => {
+    const algoliaClient = algoliaSearch(appId, apiKey)
+    return {
+      client: {
+        search(requests) {
+          return algoliaClient.search(
+            requests.map((request) => {
+              //  instant search fires an empty query on page load to ensure results are immediately available.
+              //  we exclude that result from our analytics to keep our click through rate clean
+              //  ref: https://www.algolia.com/doc/guides/getting-insights-and-analytics/search-analytics/out-of-the-box-analytics/how-to/how-to-remove-empty-search-from-analytics/
+              if (!request.params.query || request.params.query.length === 0) {
+                request.params.analytics = false
+              }
+              return request
+            })
+          )
+        },
+      },
+      initAlgoliaInsights: () => searchInsights('init', { appId, apiKey }),
+      logClick: (hit) => {
+        // @ts-expect-error - the type for this event is wrong, ref: https://github.com/algolia/search-insights.js/issues/271
+        // Fixed in 2.1.0 (cur 1.8.0)
+        return searchInsights('clickedObjectIDsAfterSearch', {
+          eventName: 'CLICK_HIT',
+          index: indexName,
+          queryID: hit.__queryID,
+          objectIDs: [hit.objectID],
+          positions: [hit.__position],
+        })
+      },
+    }
+  }, [apiKey, appId, indexName])
+
+  return (
+    <SearchContext.Provider
+      value={{
+        // Client
+        client,
+        indexName,
+        initAlgoliaInsights,
+        logClick,
+        // State
+        query,
+        setQuery,
+        isCancelled,
+        setCancelled,
+      }}
```

Comment:
So the key here is that the full context value object is memoized, otherwise it'll change referentially on each render. We'll want to include the entire object as declared here in the `useMemo` above and pass the result directly to the provider.

```tsx
const contextValue = useMemo(() => ({ ... })

<SearchContext.Provider value={contextValue}>
```

Reply:
Sorry I wasn't clear the first time around 😅

Reply:
Can do - I am curious to try to figure something out while we're at it, cause I'm not as familiar with `useMemo`.

If we're declaring the `algoliaClient` instance like `const algoliaClient = algoliaSearch(appId, apiKey)` in the `useMemo`, and `query` is a dependency of that `useMemo`, then we'll be re-running those declarations every time `query` changes, ie every time someone types something in the search bar, right?

eg

```
// Memo-ize the full context value,
// (is there any performance concern with re-declaring `algoliaClient` etc
// every time the `query` string changes?)
const contextValue = useMemo(() => (
   const algoliaClient = algoliaSearch(appId, apiKey)
    return {
      client: { ... }.
      initAlgoliaInsights: () => searchInsights('init', { appId, apiKey }),
      logClick: (hit) => { ... },
      indexName,
      query,
      setQuery,
      isCancelled,
      setCancelled,
    }
), [ apiKey, appId, indexName, query, setQuery, isCancelled, setCancelled ])
```

Maybe this is inconsequential performance wise, but wondering if it makes any sense to do eg:

```
// Memo-ize client context values, as these should only
// change and cause re-renders when Algolia configuration changes
 const { client, initAlgoliaInsights, logClick } = useMemo(() => {
    const algoliaClient = algoliaSearch(appId, apiKey)
    return {
      client: { ... }.
      initAlgoliaInsights: () => searchInsights('init', { appId, apiKey }),
      logClick: (hit) => { ... },
    }
  }, [apiKey, appId, indexName])

// Memo-ize the full context value,
// if this is not done then the `value` object would change referentially on each render
const contextValue = useMemo(() => (
    return { client, initAlgoliaInsights, logClick, indexName, query, setQuery, isCancelled, setCancelled }
), [ client, initAlgoliaInsights, logClick, indexName, query, setQuery, isCancelled, setCancelled ])
```

(Or maybe another way to put this is I don't currently understand whether the above two blocks are really that different, but would like to!)

Reply:
> If we're declaring the algoliaClient instance like const algoliaClient = algoliaSearch(appId, apiKey) in the useMemo, and query is a dependency of that useMemo, then we'll be re-running those declarations every time query changes, ie every time someone types something in the search bar, right?

That's right!

I missed that query was part of the value here, whoops. To me, your second case makes sense as those three values are relatively stable and static, whereas the others are less-so with the inclusion of `query`.

Reply:
awesome possum! Thanks for the thorough review and patience here, the objects-change-referentially thing is definitely one of those things I am really not as familiar with so kind of a big blind spot for me sometimes 😅 I started https://justjavascript.com at some point, and this thread is making me wanna go back and work all the way through it 🙇

Reply:
It's definitely an interesting concept that can have a big impact on React performance. Here's some quick links:
* https://dev.to/tylerthehaas/referential-equality-in-react-127h
* https://kentcdodds.com/blog/how-to-optimize-your-context-value